### PR TITLE
Added Industrial Protocol S7Comm

### PR DESCRIPTION
**Pull request by saffetblt**
I have to add protocol for a network project. I want to contribute.

### CODE REVIEW

**Comment by lucaderi** [comment]
Can you please update this code so I can merge it?

**Comment by saffetblt** [comment]
> Can you please update this code so I can merge it?

okay, I resolved the conflict. Can you merge it ?

File: src/include/ndpi_protocol_ids.h [hunk 1]
```diff
@@ -278,10 +278,12 @@ typedef enum {
   NDPI_PROTOCOL_WHATSAPP_FILES        = 242, /* Videos, pictures, voice messages... */
   NDPI_PROTOCOL_TARGUS_GETDATA        = 243,
   NDPI_PROTOCOL_DNP3                  = 244,
+  NDPI_PROTOCOL_104                   = 245,    
   NDPI_PROTOCOL_104                   = 245,
```
Comment (tansly) [inline src/include/ndpi_protocol_ids.h]: There is a duplicate definition here that makes the build fail.
```suggestion
```

File: src/include/ndpi_protocol_ids.h [hunk 2]
```diff
@@ -301,4 +303,4 @@ typedef enum {
 #define NDPI_PROTOCOL_NO_MASTER_PROTO    NDPI_PROTOCOL_UNKNOWN
 #define NDPI_MAX_SUPPORTED_PROTOCOLS     NDPI_LAST_IMPLEMENTED_PROTOCOL
 #define NDPI_MAX_NUM_CUSTOM_PROTOCOLS    (NDPI_NUM_BITS-NDPI_LAST_IMPLEMENTED_PROTOCOL)
-#endif
+#endif
```
Comment (tansly) [inline src/include/ndpi_protocol_ids.h]: nit: There is no newline at the end of file (other files in the repo all have one). I don't know how strict the maintainers are about this, but I personally would fix it.

File: src/lib/protocols/s7comm.c [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+                           u_int32_t *id, NDPI_PROTOCOL_BITMASK *detection_bitmask) {
+
+  ndpi_set_bitmask_protocol_detection("S7COMM", ndpi_struct, detection_bitmask, *id,
+				      NDPI_PROTOCOL_S7COMM,
+				      ndpi_search_s7comm_tcp,
+				      NDPI_SELECTION_BITMASK_PROTOCOL_V4_V6_TCP_WITH_PAYLOAD_WITHOUT_RETRANSMISSION,
+				      SAVE_DETECTION_BITMASK_AS_UNKNOWN,
+				      ADD_TO_DETECTION_BITMASK);
+  *id += 1;
+}
```
Comment (tansly) [inline src/lib/protocols/s7comm.c]: Same here.

**Review by tansly — CHANGES_REQUESTED** [review]
Hi there, great work!

I was just passing by and wanted to note these (if you haven't looked into it already).

**Review by tansly — APPROVED** [review]

File: src/include/ndpi_protocol_ids.h (hunk 2 above)
Comment (saffetblt) [reply src/include/ndpi_protocol_ids.h]: thank you i fix it

File: src/lib/protocols/s7comm.c (hunk 3 above)
Comment (saffetblt) [reply src/lib/protocols/s7comm.c]: thanks

**Comment by lucaderi** [comment]
Thank you

**Pull request closed without merging**